### PR TITLE
prov/tcp: Add multi recv support for tcp

### PR DESCRIPTION
**Pull request by vkrishna**

### CODE REVIEW

File: prov/tcp/src/tcpx_shared_ctx.c [hunk 1]
```diff
@@ -141,6 +143,8 @@ static ssize_t tcpx_srx_recvv(struct fid_ep *ep, const struct iovec *iov, void *
 
 	srx_ctx = container_of(ep, struct tcpx_rx_ctx, rx_fid);
 	assert(count <= TCPX_IOV_LIMIT);
+	assert(!(srx_ctx->op_flags & FI_MULTI_RECV) || count == 1);
+
```
Comment (shefty) [inline prov/tcp/src/tcpx_shared_ctx.c]: nit: unneeded blank line

File: prov/tcp/src/tcpx_shared_ctx.c [hunk 2]
```diff
@@ -98,6 +98,7 @@ static ssize_t tcpx_srx_recvmsg(struct fid_ep *ep, const struct fi_msg *msg,
 		goto unlock;
 	}
 
+	flags |= srx_ctx->op_flags &  FI_MULTI_RECV;
```
Comment (shefty) [inline prov/tcp/src/tcpx_shared_ctx.c]: Why is this modifying the user flags by potentially OR'ing in the mult_recv flag?

File: fabtests/Makefile.am [hunk 3]
```diff
@@ -40,6 +40,7 @@ bin_PROGRAMS = \
 	functional/fi_resmgmt_test \
 	functional/fi_rdm_atomic \
 	functional/fi_rdm_multi_recv \
+	functional/fi_msg_multi_recv \
```
Comment (shefty) [inline fabtests/Makefile.am]: I did not review these changes.  Is it possible to change the fi_rdm_multi_recv to support all EP types instead?  (And rename that test).

File: prov/tcp/src/tcpx_shared_ctx.c (hunk 2 above)
Comment (vkrishna) [reply prov/tcp/src/tcpx_shared_ctx.c]: My understanding is that if srx_ctx is created with multi_recv flag set in rx_attr->op_flags, then all the buffers posted to that srx queue should be treated as multi recv buffers. So here I am checking the op_flags and setting it in flags for this posted buffer.

Comment (shefty) [reply prov/tcp/src/tcpx_shared_ctx.c]: "Operation flags define the default flags applied to an endpoint’s data transfer operations, where a flags parameter is not available."  The 'msg' calls allow the app to override the default flags.  Without this ability, the flags could not easily be cleared.

Comment (vkrishna) [reply prov/tcp/src/tcpx_shared_ctx.c]: Got it. msg flags override attr flags, not the other way.

File: fabtests/Makefile.am (hunk 3 above)
Comment (vkrishna) [reply fabtests/Makefile.am]: That is more cleaner. I will do that.

**Comment by DougSO** [comment]
If I am understanding this design correctly, it is still using a staging buffer on each ep and copying that to the multi-receive buffer.  Is it possible to stream directly to a multi-receive buffer (if there is space) to avoid having to copy from the staging buffer to the multi-receive buffer?

**Comment by vkrishna** [comment]
@DougSO : I will add description to the commits. 

If you are talking about the multi_recv_buf in endpoint, it is not a staging buffer. It just points to the start of current message, that is being received. This is used during completion generation to let the app know where that received message is in the multi_recv bufffer. Data gets copied directly from OS sock buffers to the buffer pointed by iov.

File: fabtests/functional/multi_recv.c [hunk 4]
```diff
@@ -309,17 +309,131 @@ static int init_av(void)
...
+			goto out;
+	} else {
+		ret = init_fabric();
+		if (ret)
+			goto out;
+
+		ret = init_av();
+		if (ret)
+			goto out;
+	}
```
Comment (shefty) [inline fabtests/functional/multi_recv.c]: Other fabtests do something like this:
```
if (FI_EP_MSG)
    ft_init_fabric_cm()
else
    ft_init_fabric()
```
Why is special handling needed for MSG endpoints in this test case?  Can the common code be adjusted, if needed?

Comment (vkrishna) [reply fabtests/functional/multi_recv.c]: They differ in buffer allocation for messages (FI_OPT_SKIP_MSG_ALLOC). I will try combining it.

**Comment by shefty** [comment]
@vkrishna - are we happy with this PR?

**Comment by vkrishna** [comment]
Yes. I have few follow up changes that are not quite ready. I can open a separate PR for those.

**Pull request merged**